### ISSUE TITLE
Framework for train accuracy estimates in time series classifiers

### ISSUE THREAD

**Issue by MatthewMiddlehurst** [alan-turing-institute/sktime#727]
Several time series classification algorithms can efficiently produce estimates of accuracy (or probabilities for other metrics) on the training data. A few examples include: 

- Out of bag error in interval-based forests.
- Leave one out cross-validation leveraging the nearest neighbour classifiers and already transformed data in dictionary-based approaches.
- Removing the redundancy of transforming time series multiple times for cross-validation in algorithms such as STC and ROCKET.

While these are generally useful for evaluation classifier performance, our main usage is generating weights for the HIVE-COTE classifier.

Currently there is no framework/structure for producing these, or determining if a classifier even can produce its own estimates. As far as I am aware the only classifiers which can do so currently are a few of the dictionary-based ones I have implemented such as BOSS (method below). However, as we improve the sktime HIVE-COTE implementation, each component should be able to do this ideally. 

```python
    def _get_train_probs(self, X):
        num_inst = X.shape[0]
        results = np.zeros((num_inst, self.n_classes))
        divisor = np.ones(self.n_classes) * self.n_estimators
        for i in range(num_inst):
            sums = np.zeros(self.n_classes)

            preds = Parallel(n_jobs=self.n_jobs)(
                delayed(clf._train_predict)(
                    i,
                )
                for clf in self.classifiers
            )

            for c in preds:
                sums[self.class_dictionary.get(c, -1)] += 1

            dists = sums / divisor
            for n in range(self.n_classes):
                results[i][n] = dists[n]

        return results
```

An algorithm may have multiple methods of producing an estimate also. For example, TDE can produce both out of bag and leave one out estimates. This one is more minor (and prehaps should be handled by individual classifiers), but I will include for discussion. Perhaps an inheritable class would be the way to go about this? The following follows the functionality of how we handle this in our Java package.

```python
  class EstimateMethods(Enum):
      Default = 1
      OOB = 2
      CV = 3

  class TrainProbabilityEstimator():
      def __init__(self):
          self.estimate_method = EstimateMethods.Default
  
      def _get_train_probs(self, X):
          raise NotImplementedError("abstract method")
```

**Comment by TonyBagnall** [comment]
for context, in tsml this is controlled through the base class EnhancedAbstractClassifier
https://github.com/uea-machine-learning/tsml/blob/master/src/main/java/tsml/classifiers/EnhancedAbstractClassifier.java
not sure why we didnt just use an interface, but Im sure there were reasons. 

For python, in the first instance, I suggest that we use the existence of the method 
def _get_train_probs(self, X):
as proof that the classifier can produce the estimates, then in hive cote either use this method if it exists or cross validate if it does not. OOB is the method we should default to. Slightly less hacky way, this can be added to the Capabilities flags and tested in the unit testst, then HC can just check these. I think on reflection this is better

**Comment by fkiraly** [comment]
Very interesting thought! (somewhat late to the party here)

Personally, I would:
* distinguish methods that produce in-sample *predictions* (e.g., CV-like procedures) from methods that produce some in-sample *loss estimates* (e.g., oob error estimates)
* introduce specific methods for this, e.g., `predict_CV`, `predict_proba_CV` or `score_oob`, signature to be determined
* if multiple options exist to produce either, this should be a hyperparamter set in constructor, e.g., `predict_oob_strategy` (which has a default and can be one of multiple strings)
* if the algorithm allows arbitrary losses, this should be able to take arbitrary `sklearn` metrics for classification; these may be arguments to `score_oob` etc, or the constructor if already needed there.

Evaluation using in-sample predictions or out-of-bage estimates would be a "special" evaluation method.

What do you think, @mathhewmiddlehurst, @tonybagnall, @mloning?

**Comment by TonyBagnall** [comment]
@MatthewMiddlehurst this functionality is now present, can we close this issue?

**Comment by MatthewMiddlehurst** [comment]
I think the only thing that changed is the tag correct? The current setup is fine for my usage, but unsure how visible the functionality will be for others. Should probably change the function names to something better at least.

**Comment by fkiraly** [comment]
should this perhaps be documented with the base class then?
